mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/mps/notifications/InjectableNotifierWrapper.kt

```diff
@@ -0,0 +1,64 @@
+package org.modelix.mps.sync.mps.notifications
+
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import mu.KLogger
+import org.modelix.kotlin.utils.UnstableModelixFeature
+import java.util.Objects
+
+@UnstableModelixFeature(reason = "The new modelix MPS plugin is under construction", intendedFinalization = "2024.1")
+object InjectableNotifierWrapper {
+    private val mutex = Mutex()
+    private var lastMessage = ""
+    private var lastError: Throwable? = null
+
+    var notifier: INotifier = NoOpNotifier()
+
+    fun notifyAndLogError(message: String, error: Throwable, logger: KLogger) {
+        runWithMutex {
+            if (lastMessage != message || !errorsHaveSameOrigin(lastError, error)) {
+                lastMessage = message
+                lastError = error
+
+                logger.error(error) { message }
+                notifier.error(message)
+            }
+        }
+    }
+
+    fun notifyAndLogWarning(message: String, logger: KLogger) {
+        runWithMutex {
+            if (lastMessage != message) {
+                lastMessage = message
+                lastError = null
+
+                logger.warn { message }
+                notifier.warning(message)
+            }
+        }
+    }
+
+    fun notifyAndLogInfo(message: String, logger: KLogger) {
+        runWithMutex {
+            if (lastMessage != message) {
+                lastMessage = message
+                lastError = null
+
+                logger.info { message }
+                notifier.info(message)
+            }
+        }
+    }
+
+    private fun runWithMutex(action: () -> Unit) {
+        runBlocking {
+            mutex.withLock {
+                action()
+            }
+        }
+    }
+
+    private fun errorsHaveSameOrigin(errorOne: Throwable?, errorTwo: Throwable?) =
+        Objects.equals(errorOne?.cause?.stackTrace?.first(), errorTwo?.cause?.stackTrace?.first())
```

Comment:
Seems really wired to me that exceptions are logged multiple times, and we try to filter duplicates here.
Maybe let's not filter duplicates for now.

---

When duplicates are/become really an issue, we should clean up our logic the logs of them. 
(e.g. Throw an exception in the code and let `SyncQueue` catch them and notify the user about them. )
Just logging the same errors and deduplicating them seems to be code smell for too complicated or unstructured error handling.

Reply:
The reason for filtering the errors is, if an Exception occurs while running SyncTasks, then they bubble up via the Task dependency chain. Meaning: 
- Task A throws an Exception. This gets logged either in the Task itself, or in general [try-catch block in the SyncTaskQueue](https://github.com/modelix/modelix.mps-plugins/blob/feature/modelix-755/mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/tasks/SyncQueue.kt#L148).
- If Task B depends on the completion of Task A and Task A completed exceptionally, then Task B will also get completed exceptionally (via CompletableFuture.completeExceptionally), which may get logged in SyncTaskQueue as mentioned above. (I have to find a concrete example for this, if needed.)

Thus it's nicer to filter out the repetitions of the same exception, to avoid overloading the user with the same error message e.g. 5 times.

Reply:
Some code smells are:

* stateful notification logic stateful
*comparing Stacktraces `errorsHaveSameOrigin`

A concrete problem would be, that the stateful logic would break when exceptions from different tasks chains are logged at the same time.

> Task A throws an Exception. This gets logged either in the Task itself, or in general try-catch block in the SyncTaskQueue.
Then do not log exceptions in tasks code, let  the general try-catch block handle them.
This will also simplify the task code by not because `notifierInjector.notifyAndLogError` does not need to be called.

> If Task B depends on the completion of Task A and Task A completed exceptionally, then Task B will also get completed exceptionally (via CompletableFuture.completeExceptionally), which may get logged in SyncTaskQueue as mentioned above. 

Detect this case. For example, by saving in task B the information that it completed exceptionally because a previous task was completed exceptionally. Then you can skip logging its exception in `SyncTaskQueue`.

Reply:
So:

1. I reverted the filtering, now the user gets all messages.
2. Previously, I remembered wrong. We do not log the same error at more places (or if we do then we give more context each time), but what happens is, if an error occurrs in parallel on more threads (e.g. while uploading 3 different Models in parallel, all Models bump into the same issue) then this error will be logged and reported to the user three times, independently from each other. This is the general try-catch block in the SyncQueue class.
3. I added a `stackTrace` field to `SyncTask`, which comes in handy when we have to debug the task execution manually at runtime. Without it, just based on the `action` lambda (and its capturing groups) it's quite difficult to figure out 1:1 which piece of action was run in the `task`. This `stackTrace` is used at the moment only in manual debugging and is not logged anywhere. (The `logger.error(ex)` points us anyways to the concrete place where the error occurred.)

These changes are made in a4013d42f942d269c83af594f1236daeafb93786, and I hope they are fine for the time being.